smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java

```diff
@@ -0,0 +1,178 @@
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.utils.IoUtils;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Add config and middleware to support a service with the @httpApiKeyAuth trait.
+ */
+@SmithyInternalApi
+public final class AddHttpApiKeyAuthPlugin implements TypeScriptIntegration {
+
+    /**
+     * Plug into code generation for the client.
+     *
+     * This adds the configuration items to the client config and plugs in the
+     * middleware to operations that need it.
+     *
+     * The middleware will inject the client's configured API key into the
+     * request as defined by the @httpApiKeyAuth trait. If the trait says to
+     * put the API key into a named header, that header will be used, optionally
+     * prefixed with a scheme. If the trait says to put the API key into a named
+     * query parameter, that query parameter will be used.
+     */
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+            // Add the config if the service uses HTTP API key authorization
+            RuntimeClientPlugin.builder()
+                .inputConfig(Symbol.builder().namespace(
+                    "./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/"
+                ).name("HttpApiKeyAuthInputConfig").build())
+                .resolvedConfig(Symbol.builder().namespace(
+                    "./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/"
+                ).name("HttpApiKeyAuthResolvedConfig").build())
+                .resolveFunction(Symbol.builder().namespace(
+                    "./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/"
+                ).name("resolveHttpApiKeyAuthConfig").build())
+                .servicePredicate((m, s) -> hasHttpApiKeyAuthTrait(s) && !areAllOptionalAuthOperations(m, s))
+                .build(),
+
+            // Add the middleware to operations that use HTTP API key authorization
+            RuntimeClientPlugin.builder()
+                .pluginFunction(Symbol.builder()
+                    .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                    .name("getHttpApiKeyAuthPlugin")
+                    .build())
+                .additionalPluginFunctionParamsSupplier((m, s, o) -> new HashMap<String, Object>() {{
+                    // It's safe to do getTrait().get() because the operation predicate ensures that the trait exists
+                    // `in` and `name` are required attributes of the trait, `scheme` is optional
+                    put("in", s.getTrait(HttpApiKeyAuthTrait.class).get().getIn().toString());
+                    put("name", s.getTrait(HttpApiKeyAuthTrait.class).get().getName());
+                    put("scheme", s.getTrait(HttpApiKeyAuthTrait.class).get().getScheme().orElse(null));
```

Comment:
This map is eventually [passed to `getFunctionParametersList`](https://github.com/awslabs/smithy-typescript/blob/379611e332bfcee9a4cb9d3c328089b647de0562/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/CodegenUtils.java#L147), which looks like it'll fail when receiving a `null`. There's not a test for a trait without a `scheme`, so that's a good thing to introduce either way. Since `getScheme` returns an `Optional`, this can use an `ifPresent` chain.